Library/Homebrew/tap.rb

```diff
@@ -171,17 +171,34 @@ def core_tap?
   # @option options [Boolean] :quiet If set, suppress all output.
   def install(options = {})
     require "descriptions"
-    raise TapAlreadyTappedError, name if installed?
-    clear_cache
 
+    installed = installed?
+    full_clone = options.fetch(:full_clone, true)
     quiet = options.fetch(:quiet, false)
 
+    raise TapAlreadyTappedError, name if installed && !full_clone
+
+    if installed && !(path/".git/shallow").exist?
+      ohai "#{name} is already a full clone" unless quiet
+      return
+    end
```

Comment:
Maybe worth combining the raise above and this `if`-block, as both have the common precondition `installed`?